spring-integration-core/src/main/java/org/springframework/integration/store/SimpleMessageStore.java

```diff
@@ -255,7 +292,9 @@ public MessageGroup removeMessageFromGroup(Object groupId, Message<?> messageToR
 				SimpleMessageGroup group = this.groupIdToMessageGroup.get(groupId);
 				Assert.notNull(group, "MessageGroup for groupId '" + groupId + "' " +
 						"can not be located while attempting to remove Message from the MessageGroup");
-				group.remove(messageToRemove);
+				if (group.remove(messageToRemove)) {
+	                groupUpperBound.release();
```

Comment:
Ive looked at that pull request and I like the general idea behind it but it does not address the issues in this pull request. It will need a little more work and probably some refactoring to ensure the capacity is kept accurate. If you decrement the counter without checking if the remove actually removed anything, you can easily exceed your max capacity and run yourself out of memory. If you remove something and do not decrement you can block future adds to the message store.  I've tried to add unit tests to cover some of these cases but I know they are quite lacking.   

If I remember on monday I will attach the 2.2.6 patch that we have run in production for years. The 4.2.0 version i attached had issues due to the counter decrement issue and possibly more. I have not run my load tests against it yet. I do not have 100% confidence in this pull request until I run the load tests and even then that is only going to ensure that the code works for a delay channel.

That said, I can guarantee this code is better than the release version as the release version will not pass some of the simpler tests I added :)